src/websocket/websocket_endpoint_base.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "websocketpp/client.hpp"
+#include "../logging.h"
+#include <functional>
+#include <thread>
+#include <memory>
+namespace sensr {
+  class WebSocketEndPointBase {
+  public:
+    using MsgReceiver = std::function<void(const std::string& msg)>;
+    using ErrorReceiver = std::function<void(const std::string& err)>;
+    WebSocketEndPointBase();
+    virtual ~WebSocketEndPointBase() = default;
+
+    virtual bool Connect(const std::string &uri, const MsgReceiver& func, const ErrorReceiver& err_func) = 0;
+    virtual void Close(websocketpp::close::status::value code) = 0;
+
+  protected:
+    enum struct Status {
+      kConnecting,
+      kOpen,
+      kClosed,
+      kFailed
+    };
+    std::thread thread_;
+    websocketpp::connection_hdl connection_hdl_;
+    Status status_;
+    MsgReceiver msg_receiver_;
+    ErrorReceiver err_receiver_;
+  protected:
+    template <typename T>
+    void Init(websocketpp::client<T>& endpoint);
+    template <typename T>
+    void Fin(websocketpp::client<T>& endpoint);
+    template <typename T>
+    bool Bind(const std::shared_ptr<websocketpp::connection<T>>& con, const MsgReceiver& func, const ErrorReceiver& err_func);
+    template <typename T>
+    void Unbind(websocketpp::client<T>& endpoint, websocketpp::close::status::value code);
```

Comment:
If `T` is always the same in all function for the instance, it might be better to create a class template instead of a function template. What do you think?

Reply:
Yes, I tried to make it as class template at the 1st implementation. 
But It caused websocket_endpoint_base.h should be include in `sensr_client.h` and this results one more header file need to be installed in `/usr/include`.
I hoped to hide detail implementation as much as possible. (because `/usr/include` is a kind of public folder and anyone can tweak the code by touch the header file.)
Please understand this situation.